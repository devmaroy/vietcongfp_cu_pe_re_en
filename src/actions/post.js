// Post Actions

import database from '../firebase/firebase';
import history from '../routers/history';
import { 
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    SET_POST_LOADING,
    CLEAR_POST_LOADING,
    GET_ERRORS
} from './types';


// Get Posts
export const startGetPosts = () => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( 'posts' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const posts = [];

                snapshot.forEach( ( childSnapshot ) => {
                    posts.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch({
                    type: GET_POSTS,
                    posts
                });
    
                dispatch( clearPostLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Post
export const startGetPostById = ( id ) => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( `posts/${ id }` )
            .once( 'value' )
            .then( ( snapshot ) => {
                if ( snapshot.exists() ) {
                    const post = {
                        id: snapshot.key,
                        ...snapshot.val()
                    };
        
                    dispatch({
                        type: GET_POST,
                        post
                    });
    
                    dispatch( clearPostLoading() );
                } else {
                    // Simple redirect to non existent route
                    history.push( '/not-found' );
                }
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Post
export const startAddPost = ( postData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'posts' )
            .push( postData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_POST,
                    post: {
                        id: ref.key,
                        ...postData 
                    }
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
} ;


// Edit Post 
export const startEditPost = ( id, updates ) => {
    return ( dispatch ) => {
        return database
            .ref( `posts/${ id }` )
            .update( updates )
            .then( ( ref ) => {
                dispatch({
                    type: EDIT_POST,
                    id,
                    updates
                });
    
                history.push( `/posts/${ id }` );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Post
export const startRemovePost = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `posts/${ id }` )
            .remove()
            .then( () => {
                dispatch({
                    type: REMOVE_POST,
                    id
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Post Loading
export const setPostLoading = () => ({
    type: SET_POST_LOADING
});


// Clear Post Loading
export const clearPostLoading = () => ({
    type: CLEAR_POST_LOADING
});