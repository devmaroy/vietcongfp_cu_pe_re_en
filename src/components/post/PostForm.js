import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextFieldGroup from '../common/TextFieldGroup';
import TextAreaFieldGroup from '../common/TextAreaFieldGroup';
import Spinner from '../common/Spinner';

class PostForm extends React.Component {
    state = {
        title: this.props.post ? this.props.post.title : '',
        body: this.props.post ? this.props.post.body : '',
        createdAt: this.props.post ? this.props.post.createdAt : Date.now(),
        errors: {}
    };

    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Title
        if ( ! this.state.title ) {
            formIsValid = false;
            errors[ 'title' ] = 'Title cannot be empty';
        }

        // Body
        if ( ! this.state.body ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text cannot be empty';
        } else if ( this.state.body.length < 15 ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text must be at least 15 characters.'
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;
    
        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const postData = {
                title: this.state.title,
                body: this.state.body,
                createdAt: this.state.createdAt,
                name: this.props.user.name
            };
            
            this.props.onSubmit( postData );
        }
    }

    render() {
        const { errors } = this.state;

        return (
            <form onSubmit={ this.onSubmit }>
                <TextFieldGroup 
                    name="title"
                    placeholder="Title"
                    value={ this.state.title }
                    onChange={ this.handleChange }
                    error={ errors.title }
                />
                <TextAreaFieldGroup
                    name="body"
                    placeholder="Text"
                    value={ this.state.body }
                    onChange={ this.handleChange }
                    error={ errors.body }
                />
                <button>Save Post</button>
            </form>
        );
    };
};

PostForm.propTypes = {
    user: PropTypes.object.isRequired,
    post: PropTypes.object,
};

const mapStateToProps = ( state ) => ({
    user: state.auth.user
});

export default connect( mapStateToProps )( PostForm );