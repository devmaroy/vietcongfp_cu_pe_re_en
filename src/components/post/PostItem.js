import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { createParagraphs, limitText } from '../../utils/text-helper-functions';

export const PostItem = ( props ) => {
    const { post, isAuthenticated, showActions, isSingle } = props;
    const postBody = isSingle ? createParagraphs( post.body ) : createParagraphs( limitText( post.body ) );

    return (
        <div className="post">
            <Link to={ `/posts/${ post.id }` }>
                <h1 className="page__title">{ post.title }</h1>
            </Link>
            
            <div className="page__body">
                { postBody }
            </div>
            
            <div className="page__meta">
                <h4>Author: { post.name }</h4>
                <h5>{ moment( post.createdAt ).format('MMMM Do YYYY, h:mm:ss a') }</h5>
            </div>

            { showActions && (
                <Link to={ `/posts/${ post.id }` } className="button read-more-link">Read more</Link>
            )}

            { isAuthenticated && (
                <Link className="button edit-link" to={ `/posts/${ post.id }/edit` }>edit</Link>
            )}
        </div>
    );
};

PostItem.propTypes = {
    post: PropTypes.object.isRequired
}

PostItem.defaultProps = {
    showActions: true,
    isSingle: false
};

const mapStateToProps = ( state ) => ({
    isAuthenticated: state.auth.isAuthenticated
});

export default connect( mapStateToProps )( PostItem );