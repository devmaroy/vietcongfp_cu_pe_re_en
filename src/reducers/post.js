// Post Reducer

import {
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    SET_POST_LOADING,
    CLEAR_POST_LOADING,
    SET_COMMENT_LOADING,
    CLEAR_COMMENT_LOADING
} from '../actions/types';

const postReducerDefaultState = {
    posts: [],
    post: {},
    comments: [],
    isPostLoading: false,
    isCommentLoading: false
};

export default ( state = postReducerDefaultState, action ) => {
    switch ( action.type ) {
        
        case GET_POSTS:
            return {
                ...state,
                posts: action.posts.reverse()
            };

        case GET_POST:
            return {
                ...state,
                post: action.post
            };

        case ADD_POST:
            return {
                ...state,
                posts: [ action.post, ...state.posts ]
            };

        case EDIT_POST:
            return {
                ...state,
                posts: state.posts.map( ( post ) => {
                    if ( post.id === action.id ) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };

        case REMOVE_POST:
            return {
                ...state,
                posts: state.posts.filter( ( post ) => post.id !== action.id )
            };
        
        case GET_COMMENTS: 
            return {
                ...state,
                comments: action.comments.reverse()
            }
            
        case ADD_COMMENT:
            return {
                ...state,
                comments: [ action.comment, ...state.comments ]
            };

        case REMOVE_COMMENT:
            return {
                ...state,
                comments: state.comments.filter( ( comment ) => comment.id !== action.id )
            };

        case SET_POST_LOADING:
            return {
                ...state,
                isPostLoading: true
            };

        case SET_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: true
            };

        case CLEAR_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: false
            };

        case CLEAR_POST_LOADING:
            return {
                ...state,
                isPostLoading: false
            };
    
        default:
            return state;
    };
};