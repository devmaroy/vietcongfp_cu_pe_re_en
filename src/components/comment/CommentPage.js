import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetCommentsByPostId } from '../../actions/post';
import CommentList from './CommentList';
import AddCommentPage from '../comment/AddCommentPage';
import Spinner from '../common/Spinner';

class CommentPage extends React.Component {
    componentDidMount() {
        this.props.startGetCommentsByPostId( this.props.postId );
    }

    render() {
        const { comments, isCommentLoading } = this.props.post;
        let commentsContent;
        
        if ( comments === null || isCommentLoading ) {
            commentsContent = <Spinner />;

        } else {
            commentsContent = <CommentList comments={ comments } />;
        }

        return (
            <div>
                <AddCommentPage />
                { commentsContent }
            </div>
        );
    };
};

CommentPage.propTypes = {
    post: PropTypes.object.isRequired,
    startGetCommentsByPostId: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetCommentsByPostId: ( postId ) => dispatch( startGetCommentsByPostId( postId ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( CommentPage );