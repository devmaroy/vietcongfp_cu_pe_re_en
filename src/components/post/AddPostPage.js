import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startAddPost } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostForm from './PostForm';

class AddPostPage extends React.Component {
    onSubmit = ( postData ) => {
        this.props.startAddPost( postData );
    };

    render() {
        return (
            <div>
                <h3>Create Post</h3>
                <PostForm 
                    onSubmit={ this.onSubmit }
                />
            </div>
        )
    };
};

AddPostPage.propTypes = {
    startAddPost: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    errors: state.errors
});

const mapDispatchToProps = ( dispatch ) => ({
    startAddPost: ( postData ) => dispatch( startAddPost( postData ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( AddPostPage );