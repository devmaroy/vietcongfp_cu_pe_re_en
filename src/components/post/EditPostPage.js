import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPostById, startEditPost, startRemovePost } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostForm from './PostForm';

export class EditPostPage extends React.Component {
    componentDidMount() {
        this.props.startGetPostById( this.props.match.params.id );
    };

    onRemove = () => {
        this.props.startRemovePost( this.props.match.params.id );
    };

    onSubmit = ( updates ) => {
        this.props.startEditPost( this.props.match.params.id, updates );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postEditContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postEditContent = <Spinner />;
        } else {
           postEditContent = (
               <div>
                    <h2>Edit Post</h2>
                    <PostForm
                        onSubmit={ this.onSubmit }
                        post={ post }
                    />
                    <button onClick={ this.onRemove } className="button button--secondary">Remove</button>
               </div>
           );
        }

        return (
            <div className="content-container">
                <div className="page">
                    { postEditContent }
                </div>
            </div>
        )
    }
}

EditPostPage.propTypes = {
    startGetPostById: PropTypes.func.isRequired,
    startRemovePost: PropTypes.func.isRequired,
    startEditPost: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPostById: ( id ) => dispatch( startGetPostById( id ) ),
    startEditPost: ( id, updates ) => dispatch( startEditPost( id, updates ) ),
    startRemovePost: ( id ) => dispatch( startRemovePost( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( EditPostPage );