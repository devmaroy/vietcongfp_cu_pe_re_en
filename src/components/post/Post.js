import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPostById } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostItem from './PostItem';
import CommentPage from '../comment/CommentPage';

export class Post extends React.Component {
    componentDidMount() {
        this.props.startGetPostById( this.props.match.params.id );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postContent = <Spinner />;       
        } else {
            postContent = (
                <div>
                    <PostItem key={ post.id } post={ post } showActions={ false } isSingle={ true } />
                    <CommentPage postId={ post.id } />
                </div>
            )
        }

        return (
            <div className="content-container">
                <div className="page">
                    { postContent }
                </div>
            </div>
        );
    }
}

Post.propTypes = {
    startGetPostById: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPostById: ( id ) => dispatch( startGetPostById( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( Post );     