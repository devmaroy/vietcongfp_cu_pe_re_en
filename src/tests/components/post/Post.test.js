import React from 'react';
import { shallow } from 'enzyme';
import { Post } from '../../../components/post/Post';
import PostItem from '../../../components/post/PostItem';
import CommentPage from '../../../components/comment/CommentPage';
import posts from '../../fixtures/posts';

test( 'should render correctly Post with spinner', () => {
    const post = {
        post: {},
        isPostLoading: true
    };
    
    const wrapper = shallow(
        <Post
            startGetPostById={ () => {} }
            post={ post }
            match={ { params: { id: 1 } } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
});

test( 'should render post item and comment page', () => {
    const post = {
        post: posts[1],
        isPostLoading: false
    };

    const wrapper = shallow(
        <Post
            startGetPostById={ () => {} }
            post={ post }
            match={ { params: { id: 1 } } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( PostItem ).exists() ).toBeTruthy();
    expect( wrapper.find( CommentPage ).exists() ).toBeTruthy();
});