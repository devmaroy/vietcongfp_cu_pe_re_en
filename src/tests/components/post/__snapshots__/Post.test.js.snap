// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly Post with spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <Spinner
      alt="Loading..."
      imagePath="/images/loader.gif"
      style={
        Object {
          "display": "block",
          "margin": "60px",
          "width": "80px",
        }
      }
    />
  </div>
</div>
`;

exports[`should render post item and comment page 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <div>
      <Connect(PostItem)
        isSingle={true}
        key="2"
        post={
          Object {
            "body": "Our gallery was improved and we have now new better pictures",
            "createdAt": 1200,
            "id": "2",
            "name": "JoshTh",
            "title": "Gallery updated",
          }
        }
        showActions={false}
      />
      <Connect(CommentPage)
        postId="2"
      />
    </div>
  </div>
</div>
`;
